### PR TITLE
Stabilising `chop` lemmas

### PR DESCRIPTION
**Pull request by PetarMax**
This PR brings in `chop`-related lemmas from various engagements.

~~TODO: Tests.~~

### CODE REVIEW

File: tests/specs/functional/lemmas-spec.k [hunk 1]
```diff
@@ -601,6 +608,10 @@ module LEMMAS-SPEC
     claim <k> runLemma ( #buf ( 32 , X <<Int 248 ) ) => doneLemma( #buf ( 1 , X ) +Bytes #buf ( 31 , 0 ) ) ... </k> requires #rangeUInt (  8 , X )
     claim <k> runLemma ( #buf ( 32 , X <<Int 240 ) ) => doneLemma( #buf ( 2 , X ) +Bytes #buf ( 30 , 0 ) ) ... </k> requires #rangeUInt ( 16 , X )
 
+    claim [band-maxUInt-lt-01]: <k> runLemma ( maxUInt8   &Int X <Int    256 ) => doneLemma ( true ) ... </k> requires 0 <=Int X
+    claim [band-maxUInt-lt-02]: <k> runLemma ( maxUInt48  &Int X <Int  pow48 ) => doneLemma ( true ) ... </k> requires 0 <=Int X
+    claim [band-maxUInt-lt-03]: <k> runLemma ( maxUInt160 &Int X <Int pow160 ) => doneLemma ( true ) ... </k> requires 0 <=Int X
+
```
Comment (PetarMax) [inline tests/specs/functional/lemmas-spec.k:614]: Please ignore these, they will come in via a different PR.

File: tests/specs/functional/lemmas-spec.k [hunk 2]
```diff
@@ -404,6 +404,13 @@ module LEMMAS-SPEC
           ) => doneLemma ( false ) ... </k>
     requires #rangeUInt(256, X) andBool #rangeUInt(256, Y) andBool #rangeUInt(256, C)
 
+    claim [chop-sub-01]: <k> runLemma ( chop ( X -Int Y ) ) => doneLemma ( X -Int Y ) ... </k>
+      requires #rangeUInt(256, X) andBool #rangeUInt(256, Y) andBool Y <=Int X
+
+    claim [chop-mul-01]: <k> runLemma ( chop ( X *Int Y ) ) => doneLemma ( X *Int Y ) ... </k>
+      requires 0 <=Int X andBool X <Int 2 ^Int 96
+       andBool 0 <=Int Y andBool Y <Int 2 ^Int ( 256 -Int 96 )
```
Comment (PetarMax) [inline tests/specs/functional/lemmas-spec.k:412]: These were rules that we used in engagements, but it turns out they can be proven.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/evm-int-simplification.k [hunk 3]
```diff
@@ -91,12 +91,36 @@ module EVM-INT-SIMPLIFICATION-COMMON
...
       requires #rangeUInt(256, X) andBool #rangeUInt(256, Y)
-      [simplification]
+      [simplification, concrete(Y), comm]
+
+    rule [chop-no-overflow-add-le-l]: X:Int <=Int chop ( X +Int Y:Int )        => X +Int Y <Int pow256 requires #rangeUInt(256, X) andBool #rangeUInt(256, Y) [simplification]
+    rule [chop-no-overflow-add-le-r]: X:Int <=Int chop ( Y:Int +Int X )        => X +Int Y <Int pow256 requires #rangeUInt(256, X) andBool #rangeUInt(256, Y) [simplification]
+    rule [chop-no-overflow-mul-le-l]: X:Int <=Int chop ( X *Int Y:Int ) /Int Y => X *Int Y <Int pow256 requires #rangeUInt(256, X) andBool 0 <Int Y           [simplification]
+    rule [chop-no-overflow-mul-le-r]: X:Int <=Int chop ( Y:Int *Int X ) /Int Y => X *Int Y <Int pow256 requires #rangeUInt(256, X) andBool 0 <Int Y           [simplification]
+    rule [chop-no-overflow-mul-eq-l]: X:Int ==Int chop ( X *Int Y:Int ) /Int Y => X *Int Y <Int pow256 requires #rangeUInt(256, X) andBool 0 <Int Y           [simplification, comm]
+    rule [chop-no-overflow-mul-eq-r]: X:Int ==Int chop ( Y:Int *Int X ) /Int Y => X *Int Y <Int pow256 requires #rangeUInt(256, X) andBool 0 <Int Y           [simplification, comm]
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/evm-int-simplification.k:119]: These I have proven in Z3.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k [hunk 4]
```diff
@@ -28,13 +28,6 @@ module BITWISE-SIMPLIFICATION [symbolic]
     rule _ &Int 0 => 0 [simplification]
     rule A &Int A => A [simplification]
 
-    rule chop ( ( maxUInt48 &Int X:Int ) *Int Y:Int ) => ( maxUInt48 &Int X:Int ) *Int Y
-      requires 0 <=Int X andBool 0 <=Int Y andBool Y <=Int pow208 [simplification]
-
-    rule maxUInt8   &Int X <Int 256    => true requires 0 <=Int X [simplification, smt-lemma]
-    rule maxUInt48  &Int X <Int pow48  => true requires 0 <=Int X [simplification, smt-lemma]
-    rule maxUInt160 &Int X <Int pow160 => true requires 0 <=Int X [simplification, smt-lemma]
-
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k:37]: Please ignore these, they will come in via a different PR.

**Review by anvacaru — APPROVED** [review]

**Pull request merged**